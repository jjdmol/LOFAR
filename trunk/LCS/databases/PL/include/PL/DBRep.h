//#  DBRep.h: one line description
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_DBREP_H
#define LOFAR_PL_DBREP_H

#include <lofar_config.h>

//# Includes
#include <PL/ObjectId.h>
#include <dtl/BoundIO.h>

namespace LOFAR
{
  namespace PL
  {
    // This class is an adapter class. It transforms the data in a
    // user-defined class \c T to/from an internal representation, which is
    // guaranteed to be contiguous. DTL and ODBC demand that the data they
    // manipulate is contiguous. Hence this adapter class.
    //
    // \note This class \e must be reimplemented using full class template
    // specialization, because the Persistence Layer cannot know the exact
    // layout of the user-defined class that must be transformed.
    template <typename T> struct DBRep
    {
      // This method defines the bindings between the data members in the
      // database representation DBRep and the BoundIOs object of DTL. Please
      // note that the BoundIOs class requires that the data in the DBRep are
      // contiguous in memory.
      void bindCols(dtl::BoundIOs& cols);

    private:
      // Enforce that this class is reimplemented by defining the default
      // constructor private.
      DBRep();
    };

    // @name Full class template specializations.
    //@{

    // ObjectId is part of the metadata of a persistent object. Hence, we
    // must define a specialization.
    template<> 
    struct DBRep<ObjectId>
    {
      void bindCols(dtl::BoundIOs& cols) { cols["ObjId"] == itsOid; }
      ObjectId::oid_t itsOid;
    };
    
    //@}

  } // namespace PL

} // namespace LOFAR

#endif
