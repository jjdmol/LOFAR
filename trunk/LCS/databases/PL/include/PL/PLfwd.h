//#  PLfwd.h: Forward declarations of classes in the PL package.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_PLFWD_H
#define LOFAR_PL_PLFWD_H

// \file PLfwd.h
// Forward declarations of classes in the PL package.

#include <lofar_config.h>

namespace LOFAR
{
  namespace PL
  {
    // \addtogroup PL
    // @{

    class PersistentObject;
    class QueryObject;
    template<typename T> class Collection;
    template<typename T> class DBRep;
    template<typename T> class TPersistentObject;
    template<typename T> class DBRepHolder;

    namespace Query
    {
      class Expr;
      class ColumnExprNode;
    }

    // @}

  } // namespace PL

} // namespace LOFAR

#endif
