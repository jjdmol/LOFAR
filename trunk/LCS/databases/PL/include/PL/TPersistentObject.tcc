//#  TPersistentObject.tcc: inline implementation of persistent object class.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_TPERSISTENTOBJECT_TCC
#define LOFAR_PL_TPERSISTENTOBJECT_TCC

#include <PL/TPersistentObject.h>
#include <PL/DTLBase.h>
#include <Common/Debug.h>

namespace LOFAR
{
  namespace PL
  {

    template<typename T>
    Collection< TPersistentObject<T> > 
    TPersistentObject<T>::retrieve(const Query& query, int maxObjects)
    {
      typedef dtl::DBView< DBRep<T> >  DBViewType;

      Collection< TPersistentObject<T> > ctpo;

      DBViewType view(tableName(), BCA<T>(), query.getSql());
      typename DBViewType::select_iterator iter = view.begin();

      for (int nr = 0; iter != view.end() && nr < maxObjects; ++iter, ++nr) {
	TPersistentObject<T> tpo;
	tpo.tableName (tableName());
	tpo.fromDatabaseRep(*iter);
        // If the object T is spread among several tables we must call
        // retrieve() in order to get the data from all the tables. Otherwise,
        // we will miss the data for the "owned" POs. This really isn't very
        // efficient! However, currently there is no way to do things
        // better. We need a way to generate a better select query for this.
        if (!tpo.ownedPOs().empty()) {
          tpo.retrieve();
        }
	ctpo.add(tpo);
      }
      return ctpo;
    }


    template<typename T>
    void TPersistentObject<T>::doErase() const
    {
      typedef dtl::DBView< DBRep<ObjectId> > DBViewType;
      DBViewType view(tableName(), BCA<ObjectId>());
      DBViewType::delete_iterator iter = view;

      // setup the selection parameters
      DBRep<ObjectId> rec;
      rec.itsOid = metaData().oid()->get();

      // delete this record
      *iter = rec;

      // Once we've reached here, the update was successful.
      // Reset the meta data structure.
      metaData().reset();
    }


    template<typename T>
    void TPersistentObject<T>::doInsert() const
    {
      typedef dtl::DBView< DBRep<T> > DBViewType;

      DBViewType view(tableName(), BCA<T>());
      typename DBViewType::insert_iterator  iter = view;

      // copy info of the T to the DBRep<T> class
      DBRep<T>    rec;
      toDatabaseRep (rec);

      // save this record
      *iter = rec;

      // Once we've reached here, the insert was successful.
      // Increment the version number.
      metaData().versionNr()++;
    }


    template<typename T>
    void TPersistentObject<T>::doUpdate() const
    {
      typedef dtl::DBView< DBRep<T>, DBRep<ObjectId> > DBViewType;

      DBViewType view(tableName(), BCA<T>(), 
                      "WHERE ObjId=(?)", BPA<ObjectId>());
      typename DBViewType::update_iterator iter = view;

      // copy info of the T to the DBRep<T> class
      DBRep<T>    rec;
      toDatabaseRep (rec);

      // setup the selection parameters
      iter.Params().itsOid = rec.getOid();

      // save this record
      *iter = rec;

      // Once we've reached here, the update was successful.
      // Increment the version number.
      metaData().versionNr()++;
    }


    template<typename T>
    void TPersistentObject<T>::doRetrieve(const ObjectId& oid, bool isOwnerOid)
    {
      std::string whereClause;
      if (isOwnerOid)
        whereClause = "WHERE Owner=(?)";
      else
        whereClause = "WHERE ObjId=(?)";

      typedef dtl::DBView< DBRep<T>, DBRep<ObjectId> > DBViewType;
      DBViewType view(tableName(), BCA<T>(), whereClause, BPA<ObjectId>());
      typename DBViewType::select_iterator iter = view.begin();

      iter.Params().itsOid = oid.get();

      // We should find a match! Otherwise there's some kind of logic error.
      AssertStr(iter != view.end(), "oid=" << oid.get()
                << ", isOwnerOid=" << (isOwnerOid ? "true" : "false") );  
      fromDatabaseRep(*iter);

    }

  } // namespace PL

} // namespace LOFAR

#endif
