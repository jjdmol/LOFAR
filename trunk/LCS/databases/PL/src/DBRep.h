//#  DBRep.h: one line description
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_DBREP_H
#define LOFAR_PL_DBREP_H

#include <lofar_config.h>

//# Includes
#include <PL/ObjectId.h>
#include <PL/PersistentObject.h>

namespace LOFAR
{
  namespace PL
  {
    //# Forward Declarations

    struct DBRepOid
    {
      ObjectId::oid_t itsOid;
    };

    struct DBRepMeta
    {
      ObjectId::oid_t itsOid;
      ObjectId::oid_t itsOwnerOid;
      unsigned int    itsVersionNr;
    };

    template <typename T>
    struct DBRep : public DBRepMeta
    {
      // Enforce that this struct is specialized by defining the default
      // constructor private.
    private:
      DBRep();
    };

    template<>
    struct DBRep<ObjectId> : public DBRepOid
    {
    };

    template<>
    struct DBRep<PersistentObject::MetaData> : public DBRepMeta
    {
    };

  } // namespace PL

} // namespace LOFAR

#endif
