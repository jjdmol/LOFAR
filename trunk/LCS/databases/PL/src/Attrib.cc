//#  Attrib.cc: Global methods for converting attributes to table columns.
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include <PL/TPersistentObject.h>
#include <PL/Query/ColumnExprNode.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_iostream.h>

namespace LOFAR
{
  namespace PL
  {

    // Try to resolve the argument \a nm into a PersistentObject::Pointer. A
    // class type specifier \e must start with a '@' character; this is the
    // way that the software can discriminate a class type specifier from a
    // table column name. When \a nm is identified as a valid class type
    // specifier, we will try to find a match by comparing it against the RTTI
    // names of in the list of owned POs. If a match is found, the pointer to
    // this PersistentObject is returned, else a null pointer is returned.
    PersistentObject::Pointer resolveClassType(const PersistentObject& po,
                                               const string& nm)
    {
      PersistentObject::Pointer p;
      if (nm.empty()) return p;
      ASSERTSTR (nm[0] == '@', "Expected class type specifier in " << nm);
      string cls(nm.substr(1));
      LOG_TRACE_COND_STR ("Searching \"" << cls << "\" among ownedPOs ...");
      if (cls.empty()) return p;
      PersistentObject::POContainer::const_iterator it;
      for (it = po.ownedPOs().begin(); it != po.ownedPOs().end(); ++it) {
        LOG_TRACE_CALC_STR (typeid(**it).name() << " : match?  ");
        if (typeid(**it).name() == cls) {
          p = *it;
          LOG_TRACE_CALC_STR (" yes");
          break;
        }
        else LOG_TRACE_CALC_STR (" no");
      }
      return p;
    }


    Query::ColumnExprNode* makeAttrib(const PersistentObject& po, 
                                      const string& nm)
    {
      const PersistentObject::attribmap_t& attrMap = po.attribMap();
      PersistentObject::attribmap_t::const_iterator it;
      
      LOG_TRACE_COND_STR ("Searching for \"" << nm << "\"");
      LOG_TRACE_CALC_STR ("Showing contents of attrMap: ");
      for (it = attrMap.begin(); it != attrMap.end(); ++it) {
        LOG_TRACE_CALC_STR (" " << it->first << " = " << it->second);
      }

      // Let's do some sanity checks before we start to parse the string: it
      // is a syntax error if \a nm is an empty string, or if the first or
      // last character of \a nm is a dot (".") or a colon (":").
      if (nm.empty() || 
          nm[0] == '.' || nm[nm.size()-1] == '.' ||
          nm[0] == ':' || nm[nm.size()-1] == ':')
        THROW(QueryError, "Syntax error in specification of attribute \"" 
              << nm << "\"");
      

      // First let's try to find a match for the whole string \a nm. If the
      // whole string matches then the attribute \a nm we're looking for has a
      // table column representation in the table associated with the
      // PersistentObject po. Hence we don't need to generate a join
      // expression for the ColumnExprNode.
      if ((it = attrMap.find(nm)) != attrMap.end()) {
        LOG_TRACE_CALC_STR ("Found a match:  " << nm << " --> " << it->second);
        return new Query::ColumnExprNode(po.tableName(), it->second);
      }
      LOG_TRACE_CALC_STR ("No match for \"" << nm << "\"");

      // Now we need to search for a match using only part of the string \a
      // nm.  We will start at the end of the string and move towards the
      // beginning searching for either a dot (.) or a double colon (::). We
      // will try to find a match that resolves to a class type. If we find
      // one, we can call makeAttrib using the matching class instance as
      // first argument. The right-hand side of the string is the attribute we
      // will then look for. Hence, that is the second argument to
      // makeAttrib. An exception will be thrown when the beginning of the
      // string \a nm is reached and no match is found.
      string::size_type idx = nm.size();
      string lhs, rhs;
      while (idx-- > 0) {
        if ((idx = nm.find_last_of(":.",idx)) == string::npos) break;
        if (nm[idx] == '.') {
          if (nm[idx-1] == '.' || nm[idx-1] == ':')
            THROW(QueryError, "Syntax error in specification of attribute \"" 
                  << nm << "\"");
          lhs = nm.substr(0,idx);
          rhs = nm.substr(idx+1);
        }
        else {
          idx--;
          // Colons should appear in pairs, otherwise it's a synatx error.
          if (nm[idx] != ':' || nm[idx-1] == ':' || nm[idx-1] == '.')
            THROW(QueryError, "Syntax error in specification of attribute \"" 
                  << nm << "\"");
          lhs = nm.substr(0,idx+2);
          rhs = nm.substr(idx+2);
        }
        if ((it = attrMap.find(lhs)) == attrMap.end()) {
          LOG_TRACE_CALC_STR ("No match for \"" << lhs << "\"");
          continue;
        }
        LOG_TRACE_CALC_STR ("Found a match: " << lhs << " --> " << it->second);
        LOG_TRACE_COND_STR ("Trying to resolve class type for \"" << it->second << "\"");
        PersistentObject::Pointer p = resolveClassType(po,it->second);
        if (p)
        {
          Query::Expr left(new Query::ColumnExprNode(po.tableName(),"OBJID"));
          Query::Expr right(new Query::ColumnExprNode(p->tableName(),"OWNER"));
          Query::Expr join(left==right);
          LOG_TRACE_COND_STR ("About to call makeAttrib using \"" << rhs 
                   << "\" as second argument");
          Query::ColumnExprNode* exprNode = makeAttrib(*p, rhs);
          exprNode->addConstraint(join);
          return exprNode;
        }
      }

      // Once we've reached this point, we didn't find a matching table/column
      // name associated with the attrib \a nm. We'll throw an exception to
      // notify our caller.
      THROW(QueryError, "Failed to map attribute \"" << nm 
            << "\" onto a database column name");

    }

  } // namespace PL

} // namespace LOFAR
