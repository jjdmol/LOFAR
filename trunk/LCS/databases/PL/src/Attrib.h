//#  Attrib.h: Global methods for converting attributes to table columns.
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_ATTRIB_H
#define LOFAR_PL_ATTRIB_H

#include <lofar_config.h>

//# Includes
#include <PL/TPersistentObject.h>
#include <PL/Query/Expr.h>
#include <PL/Query/ColumnExprNode.h>
#include <string>

namespace LOFAR
{
  namespace PL
  {

    // Convert the field specifier \a nm into a column expression node that
    // represents the table name and column name plus zero or more join
    // expressions for the PersistentObject \a po.
    //
    // makeAttrib() is the actual workhorse. It will try to recursively match
    // \a nm against a map of known attributes for the PersistentObject \a
    // po. The recursion ends when there are no more subfields in the field
    // specifier \a nm (a subfield is delimited by a dot (".")). 
    //
    // For example: suppose \a nm contains the string \c
    // "foo.bar". makeAttrib() will first try to match the whole string \c
    // "foo.bar" against the map of known attributes for the PersistentObject
    // \a po. If no match is found, it'll try to match the string \c "foo"
    // against the same map. If a match is found, it'll assume that this match
    // represents a class type specifier (a class type specifier is the
    // concatenation of the '@' character and the RTTI name of the associated
    // class). Next it'll try to match this class type specifier with one of
    // the entries in the list of owned POs for PersistentObject \a po. If a
    // match is found, then makeAttrib() is called with that match as its
    // first argument and the string \c "bar" (being the right-hand-side of
    // the split field specifier \a nm) as the second argument.
    //
    // \note The map of known attributes must be created manually as part of
    // the template specialization code that must be written for each
    // TPersistentObject<T>.
    //
    // \sa PersistentObject::attribMap(), TPersistentObject<T>::theirAttribMap
    Query::ColumnExprNode* makeAttrib(const PersistentObject& po, 
                                      const std::string& nm);

    // Convert the field specifier \a nm into a table name and column name
    // plus zero or more join expressions for the specified PersistentObject
    // \a po.
    Query::Expr attrib(const PersistentObject& po, const std::string& nm)
    {
      return makeAttrib(po, nm);
    }
    
    // Convert the field specifier \a nm into a table name and column name
    // plus zero or more join expressions for a TPersistentObject of type \a
    // T.
    template<typename T>
    Query::Expr attrib(const std::string& nm)
    {
      return makeAttrib(TPersistentObject<T>(), nm);
    }

  } // namespace PL

} // namespace LOFAR

#endif
