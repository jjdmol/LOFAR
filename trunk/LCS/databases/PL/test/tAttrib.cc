//#  tAttrib.cc: program to test the makeAttrib method(s).
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include <PL/TPersistentObject.h>
#include <PL/DBRep.h>
#include <PL/Attrib.h>
#include "PO_tAttrib.h"

using namespace LOFAR::PL;
using namespace dtl;
using namespace std;

// class Z
// {
// public:
//   Z() : s("Z") {}
// private:
//   friend class TPersistentObject<Z>;
//   string s;
// };

// class Y
// {
// public:
//   Y() : s("Y") {}
// private:
//   friend class TPersistentObject<Y>;
//   string s;
//   Z z;
// };

// class X
// {
// public:
//   X() : s("X") {}
// private:
//   friend class TPersistentObject<X>;
//   string s;
// };

// class A
// {
// public:
//   A() : s("A") {}
// private:
//   friend class TPersistentObject<A>;
//   string s;
// };

// class B : public A
// {
// public:
//   B() : s("B") {}
// private:
//   friend class TPersistentObject<B>;
//   string s;
//   X x;
// };

// class C : public B
// {
// public:
//   C() : s("C") {}
// private:
//   friend class TPersistentObject<C>;
//   string s;
//   Y y;
// };


// namespace LOFAR { 

//   namespace PL {

//     template<> struct DBRep<Z> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<Z>::toDBRep(DBRep<Z>& dest) const {}
//     template<> void TPersistentObject<Z>::fromDBRep(const DBRep<Z>& src) {}
//     template<> void TPersistentObject<Z>::init()
//     { metaData().tableName() = "Z"; }
//     template<> void TPersistentObject<Z>::initAttribMap()
//     { theirAttribMap["s"] = "STR_Z"; }

//     template<> struct DBRep<Y> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<Y>::toDBRep(DBRep<Y>& dest) const {}
//     template<> void TPersistentObject<Y>::fromDBRep(const DBRep<Y>& src) {}
//     template<> void TPersistentObject<Y>::init()
//     {
//       metaData().tableName() = "Y"; 
// //       Pointer p(new TPersistentObject<Z>(data().z));
// //       p->metaData().ownerOid() = metaData().oid();
// //       ownedPOs().push_back(p);
//     }
//     template<> void TPersistentObject<Y>::initAttribMap()
//     { 
//       theirAttribMap["s"] = "STR_Y"; 
// //       theirAttribMap["z"] = 
// //         "@" + string(typeid(TPersistentObject<Z>).name()); 
//       theirAttribMap["z.s"] = "STR_Z";
//     }

//     template<> struct DBRep<X> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<X>::toDBRep(DBRep<X>& dest) const {}
//     template<> void TPersistentObject<X>::fromDBRep(const DBRep<X>& src) {}
//     template<> void TPersistentObject<X>::init()
//     { metaData().tableName() = "X"; }
//     template<> void TPersistentObject<X>::initAttribMap()
//     { theirAttribMap["s"] = "STR_X"; }

//     template<> struct DBRep<A> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<A>::toDBRep(DBRep<A>& dest) const {}
//     template<> void TPersistentObject<A>::fromDBRep(const DBRep<A>& src) {}
//     template<> void TPersistentObject<A>::init()
//     { metaData().tableName() = "A"; }
//     template<> void TPersistentObject<A>::initAttribMap()
//     { theirAttribMap["s"] = "STR_A"; }

//     template<> struct DBRep<B> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<B>::toDBRep(DBRep<B>& dest) const {}
//     template<> void TPersistentObject<B>::fromDBRep(const DBRep<B>& src) {}
//     template<> void TPersistentObject<B>::init()
//     { 
//       metaData().tableName() = "B"; 
//       {
//         Pointer p(new TPersistentObject<A>(data()));
//         p->metaData().ownerOid() = metaData().oid();
//         ownedPOs().push_back(p);
//       }
//       {
//         Pointer p(new TPersistentObject<X>(data().x));
//         p->metaData().ownerOid() = metaData().oid();
//         ownedPOs().push_back(p);
//       }
//     }
//     template<> void TPersistentObject<B>::initAttribMap() 
//     { 
//       theirAttribMap["s"] = "STR_B"; 
// //       theirAttribMap["A::"] = 
// //         "@" + string(typeid(TPersistentObject<A>).name()); 
//       theirAttribMap["A::s"] = "STR_A";
//       theirAttribMap["x"] = 
//         "@" + string(typeid(TPersistentObject<X>).name()); 
//     }

//     template<> struct DBRep<C> { void bindCols(dtl::BoundIOs&) {} };
//     template<> void TPersistentObject<C>::toDBRep(DBRep<C>& dest) const {}
//     template<> void TPersistentObject<C>::fromDBRep(const DBRep<C>& src) {}
//     template<> void TPersistentObject<C>::init()
//     {
//       Pointer p;
//       p.reset(new TPersistentObject<B>(data()));
//       p->metaData().ownerOid() = metaData().oid();
//       ownedPOs().push_back(p);
//       p.reset(new TPersistentObject<Y>(data().y));
//       p->metaData().ownerOid() = metaData().oid();
//       ownedPOs().push_back(p);
//       metaData().tableName() = "C"; 
//     }
//     template<> void TPersistentObject<C>::initAttribMap()
//     {
//       theirAttribMap["s"] = "STR_C";
//       theirAttribMap["B::"] =  
//         "@" + string(typeid(TPersistentObject<B>).name());
//       theirAttribMap["y"] = 
//         "@" + string(typeid(TPersistentObject<Y>).name());
//     }

//   }
// }
    

int main(int argc, const char* argv[])
{

  Debug::initLevels(argc, argv);

  try {

    cout << "attrib<Z>(\"s\")       = " << attrib<Z>("s")       << endl;
    cout << "attrib<Y>(\"s\")       = " << attrib<Y>("s")       << endl;
    cout << "attrib<Y>(\"z.s\")     = " << attrib<Y>("z.s")     << endl;
    cout << "attrib<X>(\"s\")       = " << attrib<X>("s")       << endl;

    cout << "attrib<A>(\"s\")       = " << attrib<A>("s")       << endl;
    cout << "attrib<B>(\"s\")       = " << attrib<B>("s")       << endl;
    cout << "attrib<B>(\"x.s\")     = " << attrib<B>("x.s")     << endl;
    cout << "attrib<B>(\"A::s\")    = " << attrib<B>("A::s")    << endl;

    cout << "attrib<C>(\"s\")       = " << attrib<C>("s")       << endl;
    cout << "attrib<C>(\"y.s\")     = " << attrib<C>("y.s")     << endl;
    cout << "attrib<C>(\"y.z.s\")   = " << attrib<C>("y.z.s")   << endl;
    cout << "attrib<C>(\"B::s\")    = " << attrib<C>("B::s")    << endl;
    cout << "attrib<C>(\"B::x.s\")  = " << attrib<C>("B::x.s")  << endl;
    cout << "attrib<C>(\"B::A::s\") = " << attrib<C>("B::A::s") << endl;

  }

  catch (LOFAR::Exception& e) {
    cerr << e << endl;
  }

  return 0;
}
