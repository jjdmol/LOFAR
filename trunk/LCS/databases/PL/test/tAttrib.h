//#  tAttrib.h: Class definitions for the test program tAttrib.cc
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_PL_TEST_TATTRIB_H
#define LOFAR_PL_TEST_TATTRIB_H

#include <PL/PLfwd.h>

/****************      User-defined classes      ****************\
|*                                                              *|
|*    Here are the user-defined classes. Note the addition of   *|
|*    the friendship declaration.                               *|
|*                                                              *|
\****************************************************************/

class Y;

class Z
{
public:
  Z() : s("Z") {}
//   void set(const string& arg) { s = arg; }
//   const string get() const { return s; }
//   string& s_ref() { return s; }
//   const string& s_ref() const { return s; }
private:
  friend class LOFAR::PL::TPersistentObject<Z>;
//   friend class LOFAR::PL::TPersistentObject<Y>;
  string s;
};

class Y
{
public:
  Y() : s("Y") {}
private:
  friend class LOFAR::PL::TPersistentObject<Y>;
  string s;
  Z z;
};

class X
{
public:
  X() : s("X") {}
private:
  friend class LOFAR::PL::TPersistentObject<X>;
  string s;
};

class A
{
public:
  A() : s("A") {}
private:
  friend class LOFAR::PL::TPersistentObject<A>;
  string s;
};

class B : public A
{
public:
  B() : s("B") {}
private:
  friend class LOFAR::PL::TPersistentObject<B>;
  string s;
  X x;
};

class C : public B
{
public:
  C() : s("C") {}
private:
  friend class LOFAR::PL::TPersistentObject<C>;
  string s;
  Y y;
};

#endif
