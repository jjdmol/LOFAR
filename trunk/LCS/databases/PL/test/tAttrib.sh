#!/bin/sh
$lofar_sharedir/runtest.sh tAttrib > tAttrib.log 2>&1
