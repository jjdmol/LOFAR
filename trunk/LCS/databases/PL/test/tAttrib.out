attrib<Z>("s")       = Z.S
attrib<Y>("s")       = Y.S
attrib<Y>("z.s")     = Z.S AND (Y.OBJID=Z.OWNER)
attrib<X>("s")       = X.S
attrib<A>("s")       = A.S
attrib<B>("s")       = B.S
attrib<B>("x.s")     = X.S AND (B.OBJID=X.OWNER)
attrib<B>("A::s")    = A.S AND (B.OBJID=A.OWNER)
attrib<C>("s")       = C.S
attrib<C>("y.s")     = Y.S AND (C.OBJID=Y.OWNER)
attrib<C>("y.z.s")   = Z.S AND ((Y.OBJID=Z.OWNER) AND (C.OBJID=Y.OWNER))
attrib<C>("B::s")    = B.S AND (C.OBJID=B.OWNER)
attrib<C>("B::x.s")  = X.S AND ((B.OBJID=X.OWNER) AND (C.OBJID=B.OWNER))
attrib<C>("B::A::s") = A.S AND ((B.OBJID=A.OWNER) AND (C.OBJID=B.OWNER))
